### ISSUE TITLE
Feature Request - detect key of wrong type passed to Map.get() or Map.remove()

### ISSUE THREAD

**Issue by gregallen** [google/error-prone#2482]
Since the signature of these methods, by design, does not constrict you to use the generic type of the key, and is just `Object`, it is easy to accidentally pass the wrong variable.

This is especially true for remove(), which in the case of cache clean-up code, might not have full test coverage, or missing/difficult to assert behaviour. This has been seen to lead to leaked memory.

**Comment by Stephan202** [comment]
This sounds like what the [`CollectionIncompatibleType`](https://errorprone.info/bugpattern/CollectionIncompatibleType) check ([source](https://github.com/google/error-prone/blob/da02fd3e4ed34e501d1b4a747f0b0a34850d6a37/core/src/main/java/com/google/errorprone/bugpatterns/collectionincompatibletype/CollectionIncompatibleType.java)) already does. Is that what you're looking for?

**Comment by cushon** [comment]
Right, that sounds like `CollectionIncompatibleType`. If you have a specific example of a bug that it should have caught that it didn't, can you share it?

```java
import java.util.Map;

class T {
  void f(Map<String, Integer> m) {
    m.get(42);
    m.remove(42);
  }
}
```

```shell
javac \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
   -J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
   -J--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED \
   -J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED \
   -XDcompilePolicy=simple \
   -processorpath error_prone_core-2.8.1-with-dependencies.jar:dataflow-errorprone-3.15.0.jar:jFormatString-3.0.0.jar \
   '-Xplugin:ErrorProne -XepDisableAllChecks -Xep:CollectionIncompatibleType:ERROR' T.java
T.java:5: error: [CollectionIncompatibleType] Argument '42' should not be passed to this method; its type int is not compatible with its collection's type argument String
    m.get(42);
         ^
    (see https://errorprone.info/bugpattern/CollectionIncompatibleType)
T.java:6: error: [CollectionIncompatibleType] Argument '42' should not be passed to this method; its type int is not compatible with its collection's type argument String
    m.remove(42);
            ^
    (see https://errorprone.info/bugpattern/CollectionIncompatibleType)
2 errors
```

**Comment by gregallen** [comment]
thanks for the quick response!

First "bug"  - that rule is a bit too cryptically (albeit accurately) named. I tried to find a suitable rule for the bug (which an IntelliJ inspection picked up) by searching the bug descriptions summary page - I was looking for "map", "key", "remove" and couldn't get a hit.

As it turns out I did already have that rule enabled, and so it does look like a miss by the rule.

This is the smallest reproducer I can make:

```java
class T {

    static class K {
    }

    void f(Map<K, Integer> m) {
        m.get(Set.of(42));
        m.remove(Set.of(42));
    }
}
```

**Comment by gregallen** [comment]
Note that in my actual bug case the type being passed to remove was the value type, so easily missed by visual inspection

**Comment by nick-someone** [comment]
As `K` is a non-final class, there could be an object that extends `K` and implements `Set`, so (as overly conservative as it sounds) there are worlds where this might be a possibly safe match:

```java
class T {

    static class K {}
    static class KSet extends K implements Set<Integer> { ... }

    void f(Map<K, Integer> m) {
        Set<Integer> foo = new KSet();
        m.get(foo);
        m.remove(foo);
    }
}
```

I presume that if you make static class `K` final, or use a non-interface concrete type in your argument (like `new HashSet<>();`), you'll get the error again.

**Comment by gregallen** [comment]
ok, wow! IntelliJ's inspection is not so generous or conservative depending on your view point.

maybe there could be an experimental version of the rule that was "stricter"?

Or, maybe a warning type rule for using non-final types as keys?
